src/vcpkg/base/diagnostics.cpp

```diff
@@ -0,0 +1,198 @@
+#include <vcpkg/base/diagnostics.h>
+#include <vcpkg/base/fmt.h>
+#include <vcpkg/base/message_sinks.h>
+
+#include <fmt/compile.h>
+
+using namespace vcpkg;
+
+namespace
+{
+    static constexpr StringLiteral ColonSpace{": "};
+
+    void append_file_prefix(std::string& target, const Optional<std::string>& maybe_origin, const TextRowCol& position)
+    {
+        // file:line:col: kind: message
+        if (auto origin = maybe_origin.get())
+        {
+            target.append(*origin);
+            if (position.row)
+            {
+                target.push_back(':');
+                fmt::format_to(std::back_inserter(target), FMT_COMPILE("{}"), position.row);
+
+                if (position.column)
+                {
+                    target.push_back(':');
+                    fmt::format_to(std::back_inserter(target), FMT_COMPILE("{}"), position.column);
+                }
+            }
+
+            target.append(ColonSpace.data(), ColonSpace.size());
+        }
+    }
+
+    void append_kind_prefix(std::string& target, DiagKind kind)
+    {
+        static constexpr StringLiteral Empty{""};
+        static constexpr const StringLiteral* prefixes[] = {
+            &Empty, &MessagePrefix, &ErrorPrefix, &WarningPrefix, &NotePrefix};
+        static_assert(std::size(prefixes) == static_cast<unsigned int>(DiagKind::COUNT));
+
+        const auto diag_index = static_cast<unsigned int>(kind);
+        if (diag_index >= static_cast<unsigned int>(DiagKind::COUNT))
+        {
+            Checks::unreachable(VCPKG_LINE_INFO);
+        }
+
+        const auto prefix = prefixes[diag_index];
+        target.append(prefix->data(), prefix->size());
+    }
+}
+
+namespace vcpkg
+{
+    void DiagnosticLine::print(MessageSink& sink) const
+    {
+        std::string buf;
+        append_file_prefix(buf, m_origin, m_position);
+        switch (m_kind)
+        {
+            case DiagKind::None:
+                // intentionally blank
+                break;
+            case DiagKind::Message: buf.append(MessagePrefix.data(), MessagePrefix.size()); break;
+            case DiagKind::Error:
+            {
+                sink.print(Color::none, buf);
+                sink.print(Color::error, "error");
+                buf.assign(ColonSpace.data(), ColonSpace.size());
+            }
+            break;
+            case DiagKind::Warning:
+            {
+                sink.print(Color::none, buf);
+                sink.print(Color::warning, "warning");
+                buf.assign(ColonSpace.data(), ColonSpace.size());
+            }
+            break;
+            case DiagKind::Note: buf.append(NotePrefix.data(), NotePrefix.size()); break;
+            default: Checks::unreachable(VCPKG_LINE_INFO);
+        }
+
+        buf.append(m_message.data());
+        buf.push_back('\n');
+        sink.print(Color::none, buf);
```

Comment:
We've gotten feedback in the past that users find our error messages annoying and/or difficult to read because their terminal is bleeding. Coloring only the 'error' or 'warning' part is consistent with what compilers do:

![gcc and clang error output showing that not the whole message is red](https://github.com/microsoft/vcpkg-tool/assets/1544943/de88ee68-41b0-4aa7-b0dd-d12de15d8f57)